File: 200.number-of-islands.cpp
/*
 * @lc app=leetcode id=200 lang=cpp
 *
 * [200] Number of Islands
 */

// @lc code=start
class Solution {
   public:
    int numIslands(vector<vector<char>>& grid) {
        int h = grid.size(), w = grid[0].size(), count = 0;
        vector<pair<int, int>> coords = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        queue<pair<int, int>> next;
        int curr_y = 0, curr_x = 0, x = 0, y = 0;

        for (int i = 0; i < h; i++) {
            for (int j = 0; j < w; j++) {
                if (grid[i][j] == '1') {
                    count++;
                    next.push({i, j});
                    while (!next.empty()) {
                        pair<int, int> curr = next.front();
                        next.pop();
                        curr_y = curr.first;
                        curr_x = curr.second;
                        if (grid[curr_y][curr_x] == '1') {
                            grid[curr_y][curr_x] = '0';
                            for (int k = 0; k < 4; k++) {
                                y = curr_y + coords[k].first;
                                x = curr_x + coords[k].second;
                                if (x >= 0 && x < w && y >= 0 && y < h && grid[y][x] > 0)
                                    next.push({curr_y + coords[k].first, curr_x + coords[k].second});
                            }
                        }
                    }
                }
            }
        }

        return count;
    }
};
// @lc code=end

/*

class Solution {
   public:
    void markIsland(vector<vector<char>>& grid, int i, int j) {
        int h = grid.size(), w = grid[0].size();
        queue<pair<int, int>> next;
        next.push({i, j});

        vector<pair<int, int>> coords = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        int count = 0;

        while (!next.empty()) {
            pair<int, int> curr = next.front();
            next.pop();
            i = curr.first;
            j = curr.second;
            if (grid[i][j] == '1') {
                // printf("mark i:%i, j: %i\n", i, j);

                grid[i][j] = '0';
                count++;
                for (int k = 0; k < 4; k++) {
                    int y = i + coords[k].first;
                    int x = j + coords[k].second;
                    if (x >= 0 && x < w && y >= 0 && y < h && grid[y][x] > 0)
                        next.push({i + coords[k].first, j + coords[k].second});
                }
            }
        }
        // printf("island size: %i\n", count);
    }

    int numIslands(vector<vector<char>>& grid) {
        int h = grid.size(), w = grid[0].size(), count = 0;
        for (int i = 0; i < h; i++) {
            for (int j = 0; j < w; j++) {
                // printf("i: %i, j: %i\n", i, j);
                if (grid[i][j] == '1') {
                    count++;
                    // printf("found new island!\n");
                    markIsland(grid, i, j);
                }
            }
        }

        return count;
    }
};
*/

File: 200.number-of-islands.py
#
# @lc app=leetcode id=200 lang=python3
#
# [200] Number of Islands
#

# @lc code=start
class Solution:
    height = 0
    width = 0

    def markIsland(self, grid: List[List[str]], i: int, j: int):
        grid[i][j] = "0"
        # adjacent_land = [(i, j)]
        adjacent_land = deque([(i, j)])

        # fill adjacent land
        # iterate through adjacent_land, mark "0" and fill adjacent_land
        while adjacent_land:
            i, j = adjacent_land.popleft()
            steps = [(-1, 0), (0, 1), (1, 0), (0, -1)]
            for di, dj in steps:
                next_i = i + di
                next_j = j + dj
                if next_i > -1 and next_i < self.height and next_j > -1 and next_j < self.width and grid[next_i][next_j] == "1":
                    grid[next_i][next_j] = "0"
                    adjacent_land.append((next_i, next_j))

    def numIslands(self, grid: List[List[str]]) -> int:
        island_count = 0
        self.height = len(grid)
        self.width = len(grid[0])
        for i in range(self.height):
            for j in range(self.width):
                if grid[i][j] == "1":
                    # found island
                    island_count += 1
                    self.markIsland(grid, i, j)

        return island_count


# @lc code=end
